src/point_cloud.rs

```diff
@@ -0,0 +1,126 @@
+// Adapted from https://github.com/sebcrozet/kiss3d/blob/v0.30.0/examples/persistent_point_cloud.rs.
+
+use std::{collections::HashMap, ops::Range};
+
+use kiss3d::camera::Camera;
+use kiss3d::context::Context;
+use kiss3d::nalgebra as na;
+use kiss3d::renderer::Renderer;
+use kiss3d::resource::{
+    AllocationType, BufferType, Effect, GPUVec, ShaderAttribute, ShaderUniform,
+};
+use na::{Matrix4, Point3};
+
+pub struct PointCloudRenderer {
+    shader: Effect,
+    pos: ShaderAttribute<Point3<f32>>,
+    color: ShaderAttribute<Point3<f32>>,
+    proj: ShaderUniform<Matrix4<f32>>,
+    view: ShaderUniform<Matrix4<f32>>,
+    colored_points: GPUVec<Point3<f32>>,
+    id_map: HashMap<String, Range<usize>>,
+    point_size: f32,
+}
+
+impl PointCloudRenderer {
+    pub fn new(point_size: f32) -> PointCloudRenderer {
+        let mut shader = Effect::new_from_str(&vertex_shader_src(point_size), FRAGMENT_SHADER_SRC);
+
+        shader.use_program();
+
+        PointCloudRenderer {
+            colored_points: GPUVec::new(Vec::new(), BufferType::Array, AllocationType::StreamDraw),
+            id_map: HashMap::new(),
+            pos: shader.get_attrib::<Point3<f32>>("position").unwrap(),
+            color: shader.get_attrib::<Point3<f32>>("color").unwrap(),
+            proj: shader.get_uniform::<Matrix4<f32>>("proj").unwrap(),
+            view: shader.get_uniform::<Matrix4<f32>>("view").unwrap(),
+            shader,
+            point_size,
+        }
+    }
+
+    pub fn insert(&mut self, id: Option<String>, points: &[[f32; 3]], colors: &[[f32; 3]]) {
+        assert_eq!(points.len(), colors.len());
+        if let Some(colored_points) = self.colored_points.data_mut() {
+            if let Some(id) = id {
+                if let Some(range) = self.id_map.remove(&id) {
+                    let len = range.end - range.start;
+                    colored_points.drain(range.clone());
+                    for r in self.id_map.values_mut() {
+                        if r.start > range.start {
+                            r.start -= len;
+                            r.end -= len;
+                        }
+                    }
+                }
+
+                let start = colored_points.len();
+                colored_points.reserve(points.len());
```

Comment:
It seems that this function is pushing two elements in L61-62.
Could you check is this correct?

```suggestion
                colored_points.reserve(points.len() * 2);
```

Reply:
Good catch. you are correct.